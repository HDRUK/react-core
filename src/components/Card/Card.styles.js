import { css } from "@emotion/react";
import { getFontSizeStyle, getComponentGlobals, getSize } from "../../theme";

export const root = () => (theme) => {
  return css`
    word-break: break-word;
    width: 100%;
    position: relative;

    ${getComponentGlobals("Card", theme)}
  `;
};

export const cardHeader = () => (theme) =>
  css`
    border-bottom: 1px solid;
    padding: ${getSize(6, theme)};

    ${getComponentGlobals("CardHeader", theme)}
  `;

export const cardBody = () => (theme) =>
  css`
    padding: ${getSize(6, theme)};
  `;

export const cardFooter = () => (theme) =>
  css`
    display: flex;
    justify-content: flex-end;
    padding: ${getSize(4, theme)};
    border-top: 1px solid;

    ${getComponentGlobals("CardFooter", theme)}
  `;
