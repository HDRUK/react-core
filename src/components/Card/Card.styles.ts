import { css, Theme } from "@emotion/react";
import { getComponentGlobals, getSize } from "../../theme";

export const root = () => (theme: Theme) => {
  return css`
    word-break: break-word;
    width: 100%;
    position: relative;

    ${getComponentGlobals("Card", theme)}
  `;
};

export const cardHeader = (theme: Theme) =>
  css`
    border-bottom: 1px solid;
    padding: ${getSize(6, theme)};

    ${getComponentGlobals("CardHeader", theme)}
  `;

export const cardBody = (theme: Theme) =>
  css`
    padding: ${getSize(6, theme)};
  `;

export const cardFooter = (theme: Theme) =>
  css`
    display: flex;
    justify-content: flex-end;
    padding: ${getSize(4, theme)};
    border-top: 1px solid;

    ${getComponentGlobals("CardFooter", theme)}
  `;
