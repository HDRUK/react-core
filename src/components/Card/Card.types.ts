import { ComponentProps } from "react";
import { ComponentStyleProps, HeightStyleProps } from "../../types";

export type CardProps = ComponentProps<"div"> &
  ComponentStyleProps &
  HeightStyleProps;
export type CardHeaderProps = ComponentProps<"div">;
export type CardBodyProps = ComponentProps<"div">;
export type CardFooterProps = ComponentProps<"div">;
