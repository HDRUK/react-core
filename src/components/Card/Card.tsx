/** @jsxImportSource @emotion/react */
import { jsx } from "@emotion/react";
import { cx } from "@emotion/css";

import * as styles from "./Card.styles";
import useCommonStyles from "../../hooks/useCommonStyles";
import { CardProps } from "./Card.types";

const Card = ({
  children,
  mt,
  mb,
  ml,
  mr,
  m,
  width,
  minWidth,
  maxWidth,
  height,
  minHeight,
  maxHeight,
  className,
  ...outerProps
}: CardProps) => {
  const commonStyles = useCommonStyles({
    mt,
    mb,
    ml,
    mr,
    m,
    width,
    minWidth,
    maxWidth,
    height,
    minHeight,
    maxHeight,
  });

  return (
    <div
      role="article"
      {...outerProps}
      className={cx(className, "ui-Card")}
      css={[commonStyles, styles.root()]}
    >
      {children}
    </div>
  );
};

export default Card;
