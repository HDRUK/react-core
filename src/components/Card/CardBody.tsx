/** @jsxImportSource @emotion/react */
import { cx } from "@emotion/css";
import * as styles from "./Card.styles";
import { CardBodyProps } from "./Card.types.js";

const CardBody = ({ className, children, ...outerProps }: CardBodyProps) => {
  return (
    <div
      {...outerProps}
      className={cx(className, "ui-CardBody")}
      css={styles.cardBody}
    >
      {children}
    </div>
  );
};

export default CardBody;
