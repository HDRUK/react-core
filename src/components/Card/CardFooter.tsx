/** @jsxImportSource @emotion/react */
import { cx } from "@emotion/css";
import * as styles from "./Card.styles";
import { CardFooterProps } from "./Card.types";

const CardFooter = ({
  className,
  children,
  ...outerProps
}: CardFooterProps) => {
  return (
    <div
      {...outerProps}
      className={cx(className, "ui-CardFooter")}
      css={styles.cardFooter}
    >
      {children}
    </div>
  );
};

export default CardFooter;
