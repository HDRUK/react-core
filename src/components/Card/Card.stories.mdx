import { useState } from "react";
import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";

import Card from "./Card";
import Button from "../Button";
import CardHeader from "./CardHeader";
import CardBody from "./CardBody";
import CardFooter from "./CardFooter";

<Meta title="Molecules/Card" component={Card} />

<Canvas>
  <Story name="Basic">
    {(props) => {
      return (
        <Card open>
          <CardBody>Body</CardBody>
        </Card>
      );
    }}
  </Story>
  <Story name="Variations">
    {(props) => {
      return (
        <>
          <Card open mb={10}>
            <CardHeader>Header</CardHeader>
            <CardBody>Body</CardBody>
            <CardFooter>
              <Button>Submit</Button>
            </CardFooter>
          </Card>
          <Card open mb={10}>
            <CardHeader>Header</CardHeader>
            <CardBody>Body</CardBody>
          </Card>
          <Card open mb={10}>
            <CardBody>Body</CardBody>
            <CardFooter>
              <Button>Submit</Button>
            </CardFooter>
          </Card>
          <Card open mb={10}>
            <CardBody>Body</CardBody>
          </Card>
        </>
      );
    }}
  </Story>
</Canvas>
