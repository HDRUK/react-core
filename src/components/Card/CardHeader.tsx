/** @jsxImportSource @emotion/react */
import { cx } from "@emotion/css";
import * as styles from "./Card.styles";
import { CardHeaderProps } from "./Card.types";

const CardHeader = ({
    className,
    children,
    ...outerProps
}: CardHeaderProps) => {
    return (
        <div
            {...outerProps}
            className={cx("ui-CardHeader", className)}
            css={styles.cardHeader}>
            {children}
        </div>
    );
};

export default CardHeader;
