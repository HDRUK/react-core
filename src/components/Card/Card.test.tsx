import { RenderResult, screen } from "@testing-library/react";
import React from "react";
import Card from "./Card";
import CardHeader from "./CardHeader";
import CardBody from "./CardBody";
import CardFooter from "./CardFooter";
import { renderWithProviders } from "../../testUtils";

let wrapper: RenderResult;

describe("Given the Card component", () => {
  describe("When it is rendered", () => {
    beforeEach(() => {
      wrapper = renderWithProviders(
        <Card>
          <CardHeader className="cardHeader">Header</CardHeader>
          <CardBody className="cardBody">Body</CardBody>
          <CardFooter className="cardFooter">Footer</CardFooter>
        </Card>
      );
    });

    it("Then should match the snapshot", async () => {
      expect(wrapper.asFragment()).toMatchSnapshot();
    });

    it("Then should have the correct header class", async () => {
      const cardHeader = await screen.findByText(/Header/);

      expect(cardHeader).toContainClass("cardHeader");
    });

    it("Then should have the correct body class", async () => {
      const cardBody = await screen.findByText(/Body/);

      expect(cardBody).toContainClass("cardBody");
    });

    it("Then should have the correct footer class", async () => {
      const cardFooter = await screen.findByText(/Footer/);

      expect(cardFooter).toContainClass("cardFooter");
    });
  });
});
