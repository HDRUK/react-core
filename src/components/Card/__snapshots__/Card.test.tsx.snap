// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Card component When it is rendered Then should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui-Card css-190dtii-Card"
    role="article"
  >
    <div
      class="cardHeader ui-CardHeader css-1t93nm7-CardHeader"
    >
      Header
    </div>
    <div
      class="cardBody ui-CardBody css-4iti4b-CardBody"
    >
      Body
    </div>
    <div
      class="cardFooter ui-CardFooter css-s03n6j-CardFooter"
    >
      Footer
    </div>
  </div>
</DocumentFragment>
`;
